edb/server/protocol/auth_ext/otc.py

```diff
@@ -0,0 +1,238 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import hashlib
+import json
+import logging
+import secrets
+import uuid
+import datetime
+from typing import Any
+
+from edb.server.protocol import execute
+
+from . import errors
+
+logger = logging.getLogger(__name__)
+
+
+def generate_code() -> str:
+    return f"{secrets.randbelow(1000000):06d}"
+
+
+def hash_code(code: str) -> bytes:
+    return hashlib.sha256(code.encode('utf-8')).digest()
+
+
+async def create(
+    db: Any, factor_id: str, ttl: datetime.timedelta
+) -> tuple[str, uuid.UUID]:
+    """
+    Create a new OneTimeCode object in the database.
+
+    Args:
+        db: Database connection
+        factor_id: The ID of the factor to associate with this code
+        ttl: Time to live for the code
+
+    Returns:
+        Tuple of (code, otc_id) where code is the plain text code and otc_id is
+        the database ID
+    """
+    code = generate_code()
+    code_hash = hash_code(code)
+    expires_at = datetime.datetime.now(datetime.timezone.utc) + ttl
+
+    r = await execute.parse_execute_json(
+        db=db,
+        query="""\
+with
+    ONE_TIME_CODE := (insert ext::auth::OneTimeCode {
+        factor := <ext::auth::Factor><uuid>$factor_id,
+        code_hash := <bytes>$code_hash,
+        expires_at := <datetime>$expires_at,
+    })
+select ONE_TIME_CODE { id };""",
+        variables={
+            "factor_id": factor_id,
+            "code_hash": code_hash,
+            "expires_at": expires_at.isoformat(),
+        },
+        cached_globally=True,
+        query_tag='gel/auth',
+    )
+
+    result_json = json.loads(r.decode())
+    if len(result_json) != 1:
+        raise errors.InvalidData("Failed to create OneTimeCode")
+
+    return code, uuid.UUID(result_json[0]["id"])
+
+
+async def verify(db: Any, factor_id: str, code: str) -> str:
+    """
+    Verify a one-time code for a given factor.
+
+    This function performs all verification operations in a single atomic query:
+    - Cleanup expired codes
+    - Check rate limits
+    - Find and validate the code
+    - Delete the code if valid
+    - Record the authentication attempt
+
+    Args:
+        db: Database connection
+        factor_id: The ID of the factor
+        code: The code to verify
+
+    Returns:
+        The OneTimeCode ID if verification succeeds
+
+    Raises:
+        NoIdentityFound: If code is invalid, expired, or max attempts exceeded
+    """
+    code_hash = hash_code(code)
+
+    # Single composed query that handles the entire verification workflow
```

Comment:
I love edgeql sometimes

Reply:
Dude, seriously. I was shocked it actually worked as expected, and had to have the tests actually prove it to me.